jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceWorkItemManager.java

```diff
@@ -0,0 +1,93 @@
+package org.jbpm.persistence.processinstance;
+
+import org.drools.core.common.InternalKnowledgeRuntime;
+import org.drools.core.process.instance.WorkItem;
+import org.drools.persistence.info.WorkItemInfo;
+import org.drools.persistence.jpa.processinstance.JPAWorkItemManager;
+import org.jbpm.process.instance.impl.WorkItemRemover;
+import org.jbpm.workflow.instance.ProcessInstanceActionQueueExecutor;
+import org.jbpm.workflow.instance.impl.queue.RemoveWorkItemAction;
+import org.kie.api.runtime.process.ProcessInstance;
+import org.kie.api.runtime.process.WorkItemHandler;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is a jBPM-specific work item manager that should be used whenever
+ * 1. jBPM is in the classpath
+ * 2. and persistence *is* being used.
+ */
+public class JPAProcessInstanceWorkItemManager extends JPAWorkItemManager implements WorkItemRemover {
+
+    private static final org.slf4j.Logger logger = LoggerFactory.getLogger(JPAProcessInstanceWorkItemManager.class);
+
+    public JPAProcessInstanceWorkItemManager(InternalKnowledgeRuntime kruntime) {
+        super(kruntime);
+    }
+
+    @Override
+    protected void signalEventAndRemoveWorkItem(ProcessInstance processInstance, String event, WorkItem workItem, WorkItemInfo workItemInfo) {
+        // process instance may have finished already
+        if (processInstance != null) {
+            if( ((org.jbpm.process.instance.ProcessInstance) processInstance).isQueueBased() ) {
+                ((ProcessInstanceActionQueueExecutor) processInstance).addNewExecutionQueueToStack( true );
+                ((ProcessInstanceActionQueueExecutor) processInstance).addProcessInstanceAction(new RemoveWorkItemAction(this, workItem.getId()));
+                processInstance.signalEvent(event, workItem);
```

Comment:
looks like the processInstance.signalEvent could be moved out from if/else as it's called in both